### ISSUE TITLE
Most scripts use /bin/bash, which not all systems have. Use /usr/bin/env instead?

### ISSUE THREAD

**Issue by xplshn** [landley/toybox#494]
Possible fixes:

- Use /usr/bin/env bash

- Do not use GNU things

Else, toybox can't be build on systems without bash, like Alpine, obviously it would be trivial to install bash but reducing dependencies (on non-POSIX things) would be great.

**Comment by oliverkwebb** [comment]
[relevant](https://github.com/landley/toybox/pull/182)

**Comment by landley** [comment]
If /bin/bash isn't a reliable absolute path, why would /usr/bin/env be a reliable absolute path?

Toybox has scripts/prereq/build.sh to produce prerequisite commands in a portable way (commit 3bbc31c78b41 has an example invocation used to build toybox outside homebrew on macos, for example). It doesn't include toysh yet because it's still in pending, but once that's promoted it should include that and you should be able to "sh scripts/make.sh". There's a planned rewrite/replacement of kconfig that should also allow .config file generation to run portably.

It's not finished yet, but that's the plan.

**Comment by xplshn** [comment]
Because not all people have installed bash at /bin/bash. And an error message that reads: `make: scripts/genconfig.sh: No such file or directory` when scripts/genconfig.sh DOES indeed exist is not quite useful for such people (me including).

/usr/bin/env should be used if you want to run under `bash` because env will lookup `bash` in the user's $PATH and run it.

For example, I have bash under `~/.local/bin` because my package manager (`bigdl`) installed it there when I ran it as a user.

**Comment by landley** [comment]
If /bin/bash can't be trusted to be at that location, why would /usr/bin/env be any different?

**Comment by oliverkwebb** [comment]
> If /bin/bash can't be trusted to be at that location, why would /usr/bin/env be any different?

Because env searches your path, doing a direct /bin/bash does not, doing
`bash scripts/make.sh` either directly or telling the makefile to do it would work I think.

**Comment by enh-google** [comment]
> Because env searches your path, doing a direct /bin/bash does not, doing

his point is "if `/bin/bash` isn't there, what reason do i have to assume `/usr/bin/env` exists?".

**Comment by oliverkwebb** [comment]
On Mon, Apr 15, 2024 at 14:25, enh-google ***@***.***(mailto:On Mon, Apr 15, 2024 at 14:25, enh-google <<a href=)> wrote:

>> Because env searches your path, doing a direct /bin/bash does not, doing
>
> his point is "if /bin/bash isn't there, what reason do i have to assume /usr/bin/env exists?".

env is a pretty standard POSIX utility that’s on most-everything that runs UNIX, bash is not

**Comment by xplshn** [comment]
> env is a pretty standard POSIX utility that’s on most-everything that runs UNIX, bash is not

Absolutely! But that's not the problem, since Landley wants to use Bash syntax, the only problem is that some people may have bash installed in other directories. For example; I wouldn't install software such as Bash into my system, instead, I put such software under ~/.local/bin or /opt/compat/*

**Comment by oliverkwebb** [comment]
The solution is to tell your makefile to run a scriptwith bash (Searches path) instead of calling the shebang.

The following diff fixes this:
```diff
@@ -14,7 +14,7 @@ all: toybox
 KCONFIG_CONFIG ?= .config

 toybox generated/unstripped/toybox: $(KCONFIG_CONFIG) *.[ch] lib/*.[ch] toys/*/*.c scripts/*.sh Config.in
-       scripts/make.sh
+       bash scripts/make.sh

 .PHONY: clean distclean baseline bloatcheck install install_flat \
        uninstall uninstall_flat tests help change \
@@ -32,7 +32,7 @@ $(KCONFIG_CONFIG): $(KCONFIG_TOP)
 $(KCONFIG_TOP): generated/Config.in generated/Config.probed
 generated/Config.probed: generated/Config.in
 generated/Config.in: toys/*/*.c scripts/genconfig.sh
-       scripts/genconfig.sh
+       bash scripts/genconfig.sh

 # Development targets
 baseline: generated/unstripped/toybox
```

**Comment by xplshn** [comment]
Its the same. However, it is better to fix the scripts shebang instead of modifying the Makefile to call them with `bash`.

**Comment by tpimh** [comment]
See previous discussion here: https://github.com/landley/toybox/issues/77
